spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2019 New Relic Corporation. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import spectator_client
+import logging
+import os
+from newrelic_telemetry_sdk import CountMetric, GaugeMetric, MetricClient
+
+
+class NewRelicMetricsService(object):
+    """A metrics service for interacting with New Relic."""
+
+    def __init__(self, spectator_helper, metric_client, tags, options):
+        self.spectator_helper = spectator_helper
+        self.metric_client = metric_client
+        self.tags = tags
+
+    def parse_metric(self, service, metric_name, metric_instance, metric_data, service_data, metric_list):
+        kind = self.spectator_helper.determine_primitive_kind(
+            metric_data["kind"])
+        for metric_data_value in metric_instance["values"]:
+            tags = self.tags.copy()
+            for tag in metric_instance["tags"]:
+                tags[tag["key"]] = tag["value"]
+            tags["applicationName"] = service_data["applicationName"]
+            tags["applicationVersion"] = service_data["applicationVersion"]
+            interval = metric_data_value["t"] - \
+                service_data["__collectStartTime"]
+            if kind == spectator_client.GAUGE_PRIMITIVE_KIND:
+                metric_list.append(GaugeMetric(
+                    name=metric_name, value=metric_data_value["v"], tags=tags))
+            else:
+                metric_list.append(CountMetric(
+                    name=metric_name, value=metric_data_value["v"], interval_ms=interval, tags=tags))
+
+    def publish_metrics(self, service_metrics):
+        metric_list = []
+        spectator_client.foreach_metric_in_service_map(
+            service_metrics, self.parse_metric, metric_list)
+        # using 1000 as a known-good size, not a theoretical maximum
+        chunk_size = 1000
+        chunks = [metric_list[i:i + chunk_size]
+                  for i in xrange(0, len(metric_list), chunk_size)]
+        for chunk in chunks:
+            response = self.metric_client.send_batch(chunk)
+            response.raise_for_status()
```

Comment:
I'm not familiar with the API but sounds like this is going to raise an exception on failure.
Is that really what you want? It could be but might not. I'm just asking.

Reply:
That's the intention, since failure would generally mean either the endpoint or the API key is misconfigured. Let me know if you have a better pattern for handling this in mind.

Reply:
Here's what that looks like in the logs:
```
12:34:46 INFO Using default metric filter for "igor"
12:34:46 INFO Using default metric transformer for "igor"
12:34:46 DEBUG Collection times 5698 (ms): {'fiat': 4, 'igor': 5665, 'clouddriver': 1, 'rosco': 8, 'kayenta': 19, 'echo': 0, 'front50': 1, 'gate': 17, 'orca': 16}
12:34:46 ERROR Traceback (most recent call last):
  File "spinnaker-monitoring-daemon/spinnaker-monitoring/server_handlers.py", line 205, in __call__
    count = service.publish_metrics(service_metric_map)
  File "spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py", line 56, in publish_metrics
    response = self.metric_client.send_batch(chunk)
  File "/home/fischer/.local/lib/python2.7/site-packages/newrelic_telemetry_sdk/client.py", line 109, in send_batch
    return self.session.post(self.url, headers=headers, data=payload)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/sessions.py", line 581, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
ConnectionError: HTTPSConnectionPool(host='abc.def', port=443): Max retries exceeded with url: /metric/v1 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f5155817d10>: Failed to establish a new connection: [Errno -2] Name or service not known',))
```

The monitoring daemon continues running after the exception is raised, so in the case of something like New Relic being down this wouldn't require a restart.

Reply:
yea, I see that the daemon catches exceptions that you throw to protect against the case I was worried about so this is fine.